### DISCUSSION TITLE
Trim unused code

### DISCUSSION THREAD

**Discussion by voidray** [AvaloniaUI/Avalonia#7543] [Q&A] [answered]
Is the "Trim unused code" feature supported for Avalonia projects? I tried it with a empty project (Avalonia project template) and publishing works (net6.0) but the application doesn't start. There are no errors, but some warnungs during publishing. When starting the app there are no errors.

Some of the console output during publish:
2>System.Data.DataSet.System.Xml.Serialization.IXmlSerializable.GetSchema(): Using member 'System.Data.DataSet.WriteXmlSchema(DataSet,XmlWriter)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataSet.GetSchema uses TypeDescriptor and XmlSerialization underneath which are not trimming safe. Members from serialized types may be trimmed if not referenced directly.
2>System.Data.DataSet.System.Xml.Serialization.IXmlSerializable.ReadXml(XmlReader): Using member 'System.Data.DataSet.ReadXmlSerializableInternal(XmlReader)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataSet.ReadXml uses XmlSerialization underneath which is not trimming safe. Members from serialized types may be trimmed if not referenced directly.
2>System.Data.DataSet.System.Xml.Serialization.IXmlSerializable.WriteXml(XmlWriter): Using member 'System.Data.DataSet.WriteXmlInternal(XmlWriter)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataSet.WriteXml uses XmlSerialization underneath which is not trimming safe. Members from serialized types may be trimmed if not referenced directly.
2>System.Data.DataTable.System.Xml.Serialization.IXmlSerializable.GetSchema(): Using member 'System.Data.DataTable.GetXmlSchema()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataTable.GetSchema uses TypeDescriptor and XmlSerialization underneath which are not trimming safe. Members from serialized types may be trimmed if not referenced directly.
2>System.Data.DataTable.System.Xml.Serialization.IXmlSerializable.ReadXml(XmlReader): Using member 'System.Data.DataTable.ReadXmlSerializableInternal(XmlReader)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataTable.ReadXml uses XmlSerialization underneath which is not trimming safe. Members from serialized types may be trimmed if not referenced directly.
2>System.Data.DataTable.System.Xml.Serialization.IXmlSerializable.WriteXml(XmlWriter): Using member 'System.Data.DataTable.WriteXmlInternal(XmlWriter)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataTable.WriteXml uses XmlSerialization underneath which is not trimming safe. Members from serialized types may be trimmed if not referenced directly.
2>SampleProject.ViewLocator.Build(Object): Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(String)'. It's not possible to guarantee the availability of the target type.
2>SampleProject.ViewLocator.Build(Object): '#0' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in call to 'System.Object System.Activator::CreateInstance(System.Type)'. The return value of method 'System.Type.GetType(String)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
2>Assembly 'Avalonia.FreeDesktop' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.Remote.Protocol' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.Win32' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.X11' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.Animation' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.Base' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.Controls' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.DesignerSupport' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.DesktopRuntime' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.Markup' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.Markup.Xaml' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.OpenGL' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Avalonia.Visuals' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'CommandLine' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Dapper' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'DynamicData' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Microsoft.Data.Sqlite' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'ReactiveUI' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Splat' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'System.Drawing.Common' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'System.Reactive' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
2>Assembly 'Tmds.DBus' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries

**Comment by maxkatz6** [comment, marked as answer]
You need yo do some changes in your csproj file so trimming can work
https://github.com/AvaloniaUI/Avalonia/issues/6892

But from your console output, it seems you have more problems with trimming and not only with Avalonia.  Like DataSet/DataTable APIs.

**Comment by voidray** [comment]
Thanks, that worked. In my case at least I only had to add `<BuiltInComInteropSupport>true</BuiltInComInteropSupport>`. In the ticket also 
`<TrimMode>copyused</TrimMode>` is listed.
Regarding the other warnings you are correct, I already added some code/libraries.
